server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SubAddressingTest.java

```diff
@@ -96,40 +97,79 @@ void shouldNotAddStorageDirectiveWhenNoRight() throws Exception {
+        MailboxSession session = mailboxManager.createSystemSession(username);
 
         // create Mailbox
         MailboxId mailboxId = mailboxManager.createMailbox(
-            MailboxPath.forUser(Username.of("recipient"), "any"), session)
-            .get();
+            MailboxPath.forUser(username, "any"), session).get();
 
-        // give right to post
-        MailboxACL.ACLCommand command = MailboxACL.command()
-            .key(MailboxACL.ANYBODY_KEY)
-            .rights(MailboxACL.Right.Post)
-            .asAddition();
+        giveRightToPost(mailboxId, session);
 
-        mailboxManager.applyRightsCommand(mailboxId, command, session);
+        Mail mail = sendSubAddressedMail(true);
+
+        AttributeName recipient = AttributeName.of("DeliveryPaths_recipient@localhost");
+        assertThat(mail.attributes().map(this::unbox))
+            .doesNotContain(Pair.of(recipient, "any"));
+    }
+
+    @Test
+    void shouldAddStorageDirectiveWhenEveryoneHasRightAndSenderIsUnknown() throws Exception {
+
+        Username username = Username.of("recipient");
+        MailboxSession session = mailboxManager.createSystemSession(username);
+
+        // create Mailbox
+        MailboxId mailboxId = mailboxManager.createMailbox(
+                MailboxPath.forUser(username, "any"), session).get();
+
+        giveRightToPost(mailboxId, session);
+        Mail mail = sendSubAddressedMail(false);
+
+        AttributeName recipient = AttributeName.of("DeliveryPaths_recipient@localhost");
+        assertThat(mail.attributes().map(this::unbox))
+            .doesNotContain(Pair.of(recipient, "any"));
+    }
+
+    @Test
+    void shouldNotAddStorageDirectiveWhenNooneHasRightAndSenderIsUnknown() throws Exception {
+
+        Username username = Username.of("recipient");
+        MailboxSession session = mailboxManager.createSystemSession(username);
+
+        // create Mailbox
+        MailboxId mailboxId = mailboxManager.createMailbox(
+            MailboxPath.forUser(username, "any"), session).get();
 
-        Mail mail = sendSubAddressedMail();
+        giveRightToPost(mailboxId, session);
+        Mail mail = sendSubAddressedMail(false);
 
         AttributeName recipient = AttributeName.of("DeliveryPaths_recipient@localhost");
         assertThat(mail.attributes().map(this::unbox))
             .doesNotContain(Pair.of(recipient, "any"));
     }
 
-    private Mail sendSubAddressedMail() throws MessagingException, IOException {
+    private void giveRightToPost(MailboxId mailboxId, MailboxSession session) throws MailboxException {
+        MailboxACL.ACLCommand command = MailboxACL.command()
+            .key(MailboxACL.ANYBODY_KEY)
+            .rights(MailboxACL.Right.Post)
+            .asAddition();
+
+        mailboxManager.applyRightsCommand(mailboxId, command, session);
+    }
+
+    private Mail sendSubAddressedMail(boolean knownSender) throws MessagingException, IOException {
```

Comment:
boolean parameter is a code smell and proves things likely need a redesign

Reply:
init of the mailet has nothing to do with sending mail. Can we move it in @BeforeEach ?

Reply:
Protip: just return a FakeMail.Builder without the sender, and let caller decide if he wants sender or not and let him build the mail then execute the mailbet on it.

Caller code:

```
Mail mail = mailBuilder().sender(...).build();
testee.service(mail);
```

Which is more readable (tests shall be as straightforward as possible and avoid ccidental indirection that obfuscate them...